File: mobile/taskmgmt_app/lib/features/tasks/domain/entities/work_task.dart
enum WorkTaskStatus { toDo, inProgress, inReview, done, cancelled }

WorkTaskStatus workTaskStatusFromString(String value) => switch (value) {
      'ToDo' => WorkTaskStatus.toDo,
      'InProgress' => WorkTaskStatus.inProgress,
      'InReview' => WorkTaskStatus.inReview,
      'Done' => WorkTaskStatus.done,
      'Cancelled' => WorkTaskStatus.cancelled,
      _ => WorkTaskStatus.toDo,
    };

String workTaskStatusToApiString(WorkTaskStatus status) => switch (status) {
      WorkTaskStatus.toDo => 'ToDo',
      WorkTaskStatus.inProgress => 'InProgress',
      WorkTaskStatus.inReview => 'InReview',
      WorkTaskStatus.done => 'Done',
      WorkTaskStatus.cancelled => 'Cancelled',
    };

String workTaskStatusLabel(WorkTaskStatus status) => switch (status) {
      WorkTaskStatus.toDo => 'Cần làm',
      WorkTaskStatus.inProgress => 'Đang làm',
      WorkTaskStatus.inReview => 'Đang xem xét',
      WorkTaskStatus.done => 'Hoàn thành',
      WorkTaskStatus.cancelled => 'Đã huỷ',
    };

class WorkTask {
  const WorkTask({
    required this.id,
    required this.title,
    required this.description,
    required this.status,
    required this.dueDateUtc,
    required this.parentTaskId,
    required this.locationId,
  });

  final String id;
  final String title;
  final String? description;
  final WorkTaskStatus status;
  final DateTime? dueDateUtc;
  final String? parentTaskId;
  final String? locationId;

  bool get isOverdue =>
      dueDateUtc != null &&
      dueDateUtc!.isBefore(DateTime.now().toUtc()) &&
      status != WorkTaskStatus.done &&
      status != WorkTaskStatus.cancelled;
}

File: backend/src/TaskMgmt.Domain/Entities/WorkTask.cs
using TaskMgmt.Domain.Common;
using TaskMgmt.Domain.Enums;

namespace TaskMgmt.Domain.Entities;

public class WorkTask : AuditableEntity
{
    public required string Title { get; set; }
    public string? Description { get; set; }
    public WorkTaskStatus Status { get; set; } = WorkTaskStatus.ToDo;
    public DateTimeOffset? DueDateUtc { get; set; }
    public bool IsActive { get; set; } = true;

    // Đánh dấu đã gửi nhắc hạn để SendDueSoonReminderJob/SendOverdueReminderJob không gửi
    // trùng lặp mỗi lần job chạy. Reset về null khi DueDateUtc đổi (xem UpdateWorkTaskCommandHandler).
    public DateTimeOffset? DueSoonReminderSentAtUtc { get; set; }
    public DateTimeOffset? OverdueReminderSentAtUtc { get; set; }

    public Guid? ParentTaskId { get; set; }
    public WorkTask? ParentTask { get; set; }
    public ICollection<WorkTask> SubTasks { get; set; } = [];

    public Guid? LocationId { get; set; }
    public Location? Location { get; set; }

    public ICollection<TaskAssignee> Assignees { get; set; } = [];
    public ICollection<Comment> Comments { get; set; } = [];
    public ICollection<Attachment> Attachments { get; set; } = [];
    public ICollection<TaskHistory> Histories { get; set; } = [];
}
